### PR TITLE
Write up the different formulations of IJKLM

### PR DESCRIPTION
**Pull request by odow**
Text still to write.

I get these timings.
```julia
  # intuitive original
  4.030615 seconds (12.99 M allocations: 485.209 MiB, 3.14% gc time, 9.25% compilation time)
  3.632798 seconds (12.64 M allocations: 462.332 MiB, 2.83% gc time)
  # intuitive tuple
  2.012132 seconds (33.92 M allocations: 1.545 GiB, 16.08% gc time, 15.50% compilation time)
  1.741992 seconds (33.68 M allocations: 1.529 GiB, 17.52% gc time)
  # fast tuple
  1.196328 seconds (33.37 M allocations: 1.512 GiB, 25.87% gc time, 26.52% compilation time)
  0.859798 seconds (33.12 M allocations: 1.495 GiB, 31.13% gc time)
  # dataframe
  0.143748 seconds (461.88 k allocations: 51.189 MiB, 14.14% gc time, 21.02% compilation time)
  0.118874 seconds (435.75 k allocations: 49.386 MiB, 19.03% gc time)
  # fast dataframe
  0.142556 seconds (245.42 k allocations: 32.769 MiB, 13.89% gc time, 21.10% compilation time)
  0.092028 seconds (219.34 k allocations: 30.968 MiB)
```

### CODE REVIEW

File: _posts/2023-17-07-gams-blog.md
Comment (odow) [inline _posts/2023-17-07-gams-blog.md:205]:
```suggestion
 * The disabling of string names
```

File: _posts/2023-17-07-gams-blog.md
Comment (odow) [inline _posts/2023-17-07-gams-blog.md:224]:
```suggestion
The answer is the nested for-loop used to
```

**Review by mlubin — COMMENTED** [review]
Thanks for putting this together. I realized at the end that I should've commented on the .jl file, not the .md file.

**Comment by odow** [comment]
Screen Shot 2023-07-16 at 8 58 54 PM

File: _posts/2023-17-07-gams-blog.md
Comment (odow) [reply _posts/2023-17-07-gams-blog.md]: Removed

File: _posts/2023-17-07-gams-blog.md
Comment (odow) [reply _posts/2023-17-07-gams-blog.md]: No 😢

File: _posts/2023-17-07-gams-blog.md
Comment (odow) [reply _posts/2023-17-07-gams-blog.md]: Added an asymptotic post out to N=3x10^3. The blogpost goes out to N=4x10^5, but I don't have time to run that on my laptop. it's a bit of a pain, because you need to run the original code to build the JSON files.

File: _posts/2023-17-07-gams-blog.md
Comment (odow) [reply _posts/2023-17-07-gams-blog.md]: Added. But I assume you'll want to tweak my wording

File: _posts/2023-17-07-gams-blog.md
Comment (odow) [reply _posts/2023-17-07-gams-blog.md]: Added something, but please edit.

File: _posts/2023-17-07-gams-blog.md
Comment (odow) [reply _posts/2023-17-07-gams-blog.md]: I don't know if we want to get into the business of running the gurobipy and pyomo models, since they could also be updated.

**Comment by odow** [comment]
It might also be worth pointing out that "fast Pyomo" is faster than "fast JuMP" because they use a different data structure.

In this case, Pyomo gets given the JKL and KLM sets as dictionaries already indexed by the first two, and not as a single list:

https://github.com/justine18/performance_experiment/blob/0aa5512e34c9041d719fa8c0763fdc892e021415/IJKLM/run_pyomo.py#L122

File: assets/tutorials/2023-07-15-gams-blog.jl
Comment (chriscoey) [inline assets/tutorials/2023-07-15-gams-blog.jl:340]:
```suggestion
# code users write inside the modeling macros (the identifiers beginning with
```

File: assets/tutorials/2023-07-15-gams-blog.jl
Comment (mlubin) [inline assets/tutorials/2023-07-15-gams-blog.jl:325]: Not just significantly better, asymptotically better.

File: _posts/2023-17-07-gams-blog.md
Comment (mlubin) [reply _posts/2023-17-07-gams-blog.md]: Ok, we don't need to rerun gurobipy and pyomo. I think it's sufficient to note that the new model is asymptotically better, which would indeed affect the qualitative conclusions made by the GAMS post.

File: assets/tutorials/2023-07-15-gams-blog.jl
Comment (blegat) [inline assets/tutorials/2023-07-15-gams-blog.jl:342]: I would highlight that this makes use of the metaprogramming ability of Julia so that the reader understand why it's not possible in Python

**Comment by blegat** [comment]
Wouldn't it be even faster with `Int` instead of `String` ? The fact that the indices are prefixed with `"i"`, `"j"` etc.. is useless since they are already in different columns so there is no risk of clash. Probably this isn't relevant here since otherwise the blog post won't be applicable for cases where the indices are not integers.

**Comment by blegat** [comment]
> It might also be worth pointing out that "fast Pyomo" is faster than "fast JuMP" because they use a different data structure.
> In this case, Pyomo gets given the JKL and KLM sets as dictionaries already indexed by the first two, and not as a single list:
> https://github.com/justine18/performance_experiment/blob/0aa5512e34c9041d719fa8c0763fdc892e021415/IJKLM/run_pyomo.py#L122

Yes, I would add it in the blog post at the end

File: _posts/2023-07-15-gams-blog.md
Comment (joehuchette) [inline _posts/2023-07-15-gams-blog.md:415]: benchmarking

File: _posts/2023-07-15-gams-blog.md
Comment (joehuchette) [inline _posts/2023-07-15-gams-blog.md:10]: maybe: “this performance difference and to present an alternative {compact, readable} modeling implementation in JuMP with much better performance”. 

Essentially I would suggest avoiding the word “formulation” as AIUI you are not creating different models for the same problem but instead only different ways of building the same model (modulo row/column permutations). I’d also make clear that we are only focusing on the JuMP implementation. Also it might be good to highlight that the best solution you present is quite readable and not overly complex.

File: assets/tutorials/2023-07-15-gams-blog.jl
Comment (jd-foster) [inline assets/tutorials/2023-07-15-gams-blog.jl:221]:
```suggestion
# Using the dataframe data structure, we can compactly write an equivalent JuMP
```

File: assets/tutorials/2023-07-15-gams-blog.jl
Comment (jd-foster) [inline assets/tutorials/2023-07-15-gams-blog.jl:327]:
```suggestion
# appropriate data structure for a model is too. It's likely that Pyomo and
```

**Review by jd-foster — COMMENTED** [review]
Pile on.

**Comment by odow** [comment]
lots of changes. Plot is now
Screen Shot 2023-07-18 at 2 12 33 PM

File: _posts/2023-07-15-gams-blog.md
Comment (joaquimg) [inline _posts/2023-07-15-gams-blog.md:328]: 2_900->2_800

Comment (odow) [reply _posts/2023-07-15-gams-blog.md:328]: No, I think I really mean 2_900?

File: assets/tutorials/2023-07-15-gams-blog.jl
Comment (mlubin) [inline assets/tutorials/2023-07-15-gams-blog.jl:275]: Intuitive and fast should have quotes around them here too :)

File: assets/tutorials/2023-07-15-gams-blog.jl
Comment (mlubin) [inline assets/tutorials/2023-07-15-gams-blog.jl:83]:
```suggestion
# A typical reason for poor performance in Julia code is type instability,
```

File: assets/tutorials/2023-07-15-gams-blog.jl
Comment (mlubin) [inline assets/tutorials/2023-07-15-gams-blog.jl:86]:
```suggestion
# You can check a function for type stability using
```

File: assets/tutorials/2023-07-15-gams-blog.jl
Comment (mlubin) [inline assets/tutorials/2023-07-15-gams-blog.jl:294]: Suggested replacement paragraph:

The dataframe formulation is asymptotically faster. Once we understand that the bottleneck in this benchmark is equivalent to an inner join, it is not difficult to address it, given that general-purpose languages like Julia and Python have libraries specialized for this task. Pyomo and Gurobipy would likely benefit from a similar optimization.

Comment (blegat) [reply assets/tutorials/2023-07-15-gams-blog.jl:294]: "Once we understand" -> "Once one understands" ?

File: _posts/2023-07-15-gams-blog.md
Comment (joaquimg) [reply _posts/2023-07-15-gams-blog.md:316]: You are correct! 2900 is right.
I skipped a number XD

**Comment by odow** [comment]
Okay, I think we're pretty much there.

image

File: _posts/2023-07-15-gams-blog.md
Comment (odow) [inline _posts/2023-07-15-gams-blog.md:5]:
```suggestion
```

File: _posts/2023-07-15-gams-blog.md
Comment (mlubin) [inline _posts/2023-07-15-gams-blog.md:9]: Suggested edits:
A recent blog post [...] on a model they call IJKLM. We respond to this blog post by explaining the difference in performance and presenting an alternative JuMP implementation with asymptotically better performance. We also identify that differences in the input data format—not anything intrinsic to the respective libraries—explain the performance difference between the Pyomo model and the JuMP model. Finally, we respond to general claims about domain specific languages and the trade-offs of developing optimization models within a fully-featured programming language like Julia or Python.

**Review by mlubin — APPROVED** [review]
LGTM besides the intro paragraph.

**Comment by odow** [comment]
This is good to go by me, but I'll wait for some more approvals before merging.

**Review by joehuchette — APPROVED** [review]

File: _posts/2023-07-20-gams-blog.md
Comment (odow) [reply _posts/2023-07-20-gams-blog.md:13]: Pyomo. Related to this image:

image

@RichardOberdieck was wondering https://oberdieck.dk/p/not-so-fast-gams/

**Comment by RichardOberdieck** [comment]
Uh, nice, I guessed that you guys would follow up on this :) These timings look nice, go JuMP!

**Comment by odow** [comment]
> I have tried to “beat” GAMS two or three times in model generation speed using gurobipy, and it is pretty tough. The best I’ve gotten was around 10% slower than GAMS.

@RichardOberdieck what was the fastest? Did you try a `pandas` approach, or just using the `gp.tuplelist`? It's a bit of a weird model for a benchmark because the choice of input data structure has such a large impact.

**Review by mlubin — APPROVED** [review]

**Comment by RichardOberdieck** [comment]
So I didn't try it for this one because I don't access to a GAMS license anymore. When I was still working at Gurobi, I had a client using GAMS and I wanted to show the `gurobipy` could be as fast. So  spent about 2-3 days working on formulating that MIP in the most vectorized, and optimized form I could think of. I ended up being 10% slower roughly. The same happened with a different client a few months later, but that was a smaller model.

With `gurobipy` you can be really fast when you have matrix-like operations. There you can use the matrix API, which is very efficient.

Regarding this benchmark I agree: it is pretty useless, and I don't think anyone sense-checked the numbers. Otherwise pyomo would never have been faster than JuMP...

By the way: what is the plan now with this PR, are you going to make a blog post about it? Would be a fun read :)

**Review by blegat — APPROVED** [review]

**Comment by blegat** [comment]
> By the way: what is the plan now with this PR, are you going to make a blog post about it? Would be a fun read :)

Yes, once this PR is merged, this should appear in the blog post: https://jump.dev/pages/news/

**Comment by odow** [comment]
Here's the post: https://jump.dev/2023/07/20/gams-blog/

**Review by joaquimg — APPROVED** [review]

**Pull request merged**
